crates/ol/mempool/src/state.rs

```diff
@@ -0,0 +1,1589 @@
+//! Mempool service state management.
+
+use std::{
+    collections::{BTreeMap, BTreeSet, HashMap},
+    sync::Arc,
+};
+
+use ssz::Decode;
+use strata_acct_types::AccountId;
+use strata_db_types::types::MempoolTxData;
+use strata_identifiers::{OLBlockCommitment, OLTxId};
+use strata_ledger_types::{IAccountState, ISnarkAccountState, IStateAccessor};
+use strata_ol_chain_types_new::{OLBlock, OLTransaction, TransactionPayload};
+use strata_ol_state_types::OLState;
+use strata_service::ServiceState;
+use strata_storage::NodeStorage;
+
+use crate::{
+    OLMempoolError, OLMempoolResult,
+    ordering::{FifoOrderingStrategy, OrderingStrategy},
+    types::{
+        MempoolEntry, MempoolOrderingKey, OLMempoolConfig, OLMempoolRejectReason, OLMempoolStats,
+        OLMempoolTransaction, OLMempoolTxPayload,
+    },
+    validation::validate_transaction,
+};
+
+/// Immutable context for mempool service (shared via Arc).
+#[derive(Clone)]
+pub(crate) struct MempoolContext {
+    /// Mempool configuration.
+    pub(crate) config: OLMempoolConfig,
+
+    /// Storage backend for creating StateAccessor instances.
+    /// Mempool creates StateAccessor per-operation based on current_tip.
+    pub(crate) storage: Arc<NodeStorage>,
+
+    /// Ordering strategy for transaction priority.
+    pub(crate) ordering_strategy: Arc<dyn OrderingStrategy>,
+}
+
+impl std::fmt::Debug for MempoolContext {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("MempoolContext")
+            .field("config", &self.config)
+            .field("ordering_strategy", &"<OrderingStrategy>")
+            .field("storage", &"<NodeStorage>")
+            .finish()
+    }
+}
+
+/// Combined state for the service (context + mutable state).
+#[derive(Debug)]
+pub(crate) struct MempoolServiceState {
+    ctx: Arc<MempoolContext>,
+
+    /// In-memory entries indexed by transaction ID.
+    entries: HashMap<OLTxId, MempoolEntry>,
+
+    /// Ordering index: priority → transaction ID.
+    /// BTreeMap provides ordered iteration by priority (lower = higher priority).
+    /// Priority is u128 encoding (primary_key << 64) | insertion_id.
+    ordering_index: BTreeMap<u128, OLTxId>,
+
+    /// Per-account transaction index for cascade removal.
+    /// Maps account → set of transaction IDs for that account.
+    /// Used to find and remove subsequent transactions when one fails.
+    by_account: HashMap<AccountId, BTreeSet<OLTxId>>,
+
+    /// Next insertion ID for deterministic ordering.
+    next_insertion_id: u64,
+
+    /// Current chain tip (slot + block ID) for timestamping, expiry, and reorg handling.
+    current_tip: OLBlockCommitment,
+
+    /// Track expected sequence number per account (for gap detection).
+    /// For each account, stores the next expected seq_no (current + pending count).
```

Comment:
just checking, this is _only_ for accounts that currently have a mempool tx targeting them, right?